File: 121.best-time-to-buy-and-sell-stock.ts
/*
 * @lc app=leetcode id=121 lang=typescript
 *
 * [121] Best Time to Buy and Sell Stock
 */

// @lc code=start
function maxProfit(prices: number[]): number {
   let max = 0;
   let buy = 0;
   const length = prices.length;

   for (let i = 0; i < length; i++) {
      if (prices[i] < prices[buy]) {
         buy = i;
      } else {
         const profit = prices[i] - prices[buy];
         if (profit > max) {
            max = profit;
         }
      }
   }

   return max;
}
// @lc code=end

File: 121.best-time-to-buy-and-sell-stock.py
#
# @lc app=leetcode id=121 lang=python3
#
# [121] Best Time to Buy and Sell Stock
#

# @lc code=start
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        buyPrice = prices[0]
        maxProfit = 0
        
        for i in range(1, len(prices)):
            profit = prices[i] - buyPrice
            #print("buy price:", buyPrice)
            #print("sell price:", prices[i])
            maxProfit = max(maxProfit, profit)
            buyPrice = min(buyPrice, prices[i])
            
        return maxProfit
            
# @lc code=end

